New module - meraki_network

kbreit opened the pull request:
##### SUMMARY
New module for Meraki suite of modules. `meraki_network` is responsible for CRUD functionality related to networks. A relatively full suite of integration tests are included in this pull request.

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
meraki_network

##### ANSIBLE VERSION
```
ansible 2.6.0 (meraki/meraki_net a219509795) last updated 2018/05/06 20:45:34 (GMT -500)
  config file = None
  configured module search path = ['/Users/kbreit/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/kbreit/Documents/Programming/ansible/lib/ansible
  executable location = /Users/kbreit/Documents/Programming/ansible/bin/ansible
  python version = 3.5.4 (default, Feb 25 2018, 14:56:02) [GCC 4.2.1 Compatible Apple LLVM 9.0.0 (clang-900.0.39.2)]
```

##### ADDITIONAL INFORMATION
Functions supported:
- Query all networks
- Query one network
- Create network
- Update network with idempotency
- Delete network

kbreit committed:
New module for CRUD functionality of networks in a Meraki environment
- Relatively full integration test suite
- More functions to come

kbreit committed:
Fix indent for PEP8.

- Look into why this didn't show on a local PEP8 test

dagwieers commented on lib/ansible/module_utils/network/meraki/meraki.py:
```diff
@@ -76,7 +76,7 @@ def __init__(self, module, function=None):
         # If URLs need to be modified or added for specific purposes, use .update() on the url_catalog dictionary
         self.get_urls = {'organizations': '/organizations',
-                         'networks': '/organizations/{org_id}/networks',
+                         'network': '/organizations/{org_id}/networks',
```
Not sure why this is singular now ?

kbreit replied:
Consistency. The module is network. I set networks early on when I first wrote the module, before meraki_network was a thing.

dagwieers replied:
Ok, so you're standardizing on the module name, not the internal REST call.
I guess both are fine, as it's not exactly exposed to the user so it doesn't really matter.

dagwieers commented on lib/ansible/module_utils/network/meraki/meraki.py:
```diff
@@ -184,16 +184,30 @@ def get_org_id(self, org_name):
         if not data:
             org_id = self.get_org_id(org_name)
-            path = '/organizations/{org_id}/networks/{net_id}'.format(org_id=org_id, net_id=self.get_net_id(org_name=org_name, net_name=net_name, data=data))
+            path = '/organizations/{org_id}/networks/{net_id}'.format(
```
I think this would be more readable if you do:

```python
    org_id = self.get_org_id(org_name)
    net_id = self.get_net_id(org_name, net_name, data)
    path = '/organizations/{org_id}/networks/{net_id}'.format(org_id=org_id, net_id=net_id)
```

So using positional arguments where there's no discussion, moving different operations in separate steps, and bring the information together at the very end. The path-string could be coming from somewhere else (if this is something that is being reused, but probably not relevant here).

kbreit replied:
That's correct. I always thought REST endpoints, by best practice, were supposed to be singular and not plural. So the Meraki API doesn't abide by that and calling it meraki_networks and meraki_admins and whatnot seems awkward to me.

dagwieers replied:
No, no, the module name should definitely be singular.

I would probably use internally the same name as the REST API, especially because it is not exposed to the user and the internal stuff is in itself more consistent. But again, a personal choice, not something that is written into stone.

(I think the plurar in the REST interface does make sense and is a common thing even, because it does e.q. affect networks, but that is debatable).

dagwieers commented on lib/ansible/module_utils/network/meraki/meraki.py:
```diff
@@ -236,7 +250,9 @@ def request(self, path, method=None, payload=None):
         if self.status >= 300:
-            self.fail_json(msg='Request failed for {url}: {status} - {msg}'.format(**info))
+            if not payload:
+                self.fail_json(msg='Request failed for {url}: {status} - {msg}'.format(**info))
```
I don't see the point here, both errors are identical, so why do this ?
If it fails you wouldn't even know which one actually failed.

dagwieers commented on lib/ansible/modules/network/meraki/meraki_network.py:
```diff
@@ -0,0 +1,259 @@
+    state:
+        description:
+        - Create or modify an organization.
+        choices: ['present', 'query', 'absent']
```
So, my personal view here is that if you list items, and the order has no real meaning, it's better to order them alphabetically.

dagwieers replied:
There's also no need to quote the values here.
Another rule I abide to is to not quote strings that not need quoting in YAML.
Otherwise people will not understand the YAML rules to quoting, and start quoting everything always.

dagwieers reviewed (approved):
LGTM

kbreit committed:
Dag requested changes.
- Removed a section in get_net as its backend isn't implemented
- Documentation modifications

dagwieers commented on lib/ansible/module_utils/network/meraki/meraki.py:
```diff
@@ -184,16 +184,30 @@ def get_org_id(self, org_name):
-            for n in data:
-                if n['name'] == net_name:
-                    return n
+        # if not data:
```
Why is this now commented out ?

kbreit replied:
That had some additional functionality in it if you don't pass `data`. But I didn't have the backend fully implemented so I removed the functionality for now.

dagwieers commented on lib/ansible/modules/network/meraki/meraki_network.py:
```diff
@@ -0,0 +1,259 @@
+'''
+
+import os
+from ansible.module_utils.basic import AnsibleModule, json, env_fallback
```
We tend to sort the imports by name, but also by module.

dagwieers reviewed (approved):
LGTM
